ingest/entry/entry.go

```diff
@@ -45,108 +46,238 @@ type Entry struct {
+// we assume the caller has already ensured that the buffer is large enough to at least contain a header
+func DecodeHeaderTagSec(buff []byte) (tag EntryTag, sec int64) {
+	tag = EntryTag(binary.LittleEndian.Uint16(buff[16:]))
+	sec = int64(binary.LittleEndian.Uint64(buff[4:]))
+	return
+}
+
+// EntrySize just decodes enough of the header to decide the actual encoded size of an entry
+// this function is typically used for rapidly skipping an entry
+func EntrySize(buff []byte) (n int, err error) {
+	if len(buff) < ENTRY_HEADER_SIZE {
+		err = ErrInvalidHeader
+		return
 	}
-	ent.TS.Decode(buff[4:])
-	ent.Tag = EntryTag(binary.LittleEndian.Uint16(buff[16:]))
-	if ipv4 {
-		ent.SRC = buff[18:22]
-	} else {
-		ent.SRC = buff[18:ENTRY_HEADER_SIZE]
+	datasize := binary.LittleEndian.Uint32(buff)
+	flags := uint8(datasize >> 30)
+
+	n = int(datasize & flagMask) // clear flags from datasize
+	if len(buff) < n {
+		err = ErrInvalidBufferSize
+		return
+	}
+	if (flags & flagEVs) == 0 {
+		return
+	}
+
+	//we have EVs, check the buffer again
+	var hdr EVBlockHeader
+	if hdr, err = DecodeEVBlockHeader(buff[n:]); err == nil {
+		n += int(hdr.Size)
 	}
-	return int(datasize)
+	return
+}
+
+// DecodePartialHeader decodes only the timestamp second, tag, hasEvs, and DataSize
+// this function is used for quickly scanning through entries in their encoded form
+func DecodePartialHeader(buff []byte) (ts Timestamp, tag EntryTag, ipv4, hasEvs bool, datasize uint32) {
+	//decode the datasize and grab the flags from the datasize
+	datasize = binary.LittleEndian.Uint32(buff)
+	flags := uint8(datasize >> 30)
+	datasize &= flagMask // clear flags from datasize
+	hasEvs = ((flags & flagEVs) != 0)
+	ipv4 = ((flags & flagIPv4) != 0)
+	tag = EntryTag(binary.LittleEndian.Uint16(buff[16:]))
+	ts.Decode(buff[4:])
+	return
+}
+
+// decodeHeader copies copies the SRC buffer
+func (ent *Entry) decodeHeader(buff []byte) (int, bool) {
+	var hasEvs bool
+	var datasize uint32
+	var src net.IP
+	ent.TS, src, ent.Tag, hasEvs, datasize = DecodeHeader(buff)
+	ent.SRC = append(net.IP(nil), src...)
+	return int(datasize), hasEvs
 }
 
-func (ent *Entry) DecodeHeader(buff []byte) (int, error) {
+// decodeHeaderAlt gets a direct handle on the SRC buffer
+func (ent *Entry) decodeHeaderAlt(buff []byte) (int, bool) {
+	var hasEvs bool
+	var datasize uint32
+	ent.TS, ent.SRC, ent.Tag, hasEvs, datasize = DecodeHeader(buff)
+	return int(datasize), hasEvs
+}
+
+func (ent *Entry) DecodeHeader(buff []byte) (int, bool, error) {
 	if len(buff) < ENTRY_HEADER_SIZE {
-		return 0, ErrInvalidBufferSize
+		return 0, false, ErrInvalidBufferSize
 	}
-	return ent.decodeHeader(buff), nil
+	dataLen, hasEvs := ent.decodeHeader(buff)
+	return dataLen, hasEvs, nil
 }
 
 // DecodeEntry will copy values out of the buffer to generate an entry with its own
 // copies of data.  This ensures that entries don't maintain ties to blocks
 // DecodeEntry assumes that a size check has already happened
```

Comment:
Remember that the docs ignores a single newline, so you need to end sentences with periods.